File: backend/src/handlers/reminder.ts
import { isValidObjectId } from 'mongoose';
import { expectRecord, readTrimmedString } from '../http/body';
import { toApiFailure } from '../http/errors';
import { ApiEnvelope, fail, success } from '../http/response';
import { Reminder } from '../models/reminder';
import {
  ReminderListPayload,
  ReminderPlanRecord,
  TodayReminderItemRecord,
  TodayRemindersPayload,
} from '../types';

type ReminderMedicineRefLike = {
  drugCode: string;
  approvalNo: string;
  productName: string;
};

type ReminderDocLike = {
  _id: unknown;
  userId?: string;
  time?: string;
  drugCode?: string;
  approvalNo?: string;
  productName?: string;
  medicines?: unknown;
  dosage?: string;
  subtitle?: string;
  enabled?: boolean;
  repeatRule?: string;
  method?: string;
  startDate?: string;
  endDate?: string;
};

const timeRegExp = /^\d{2}:\d{2}$/;
const dateRegExp = /^\d{4}-\d{2}-\d{2}$/;

function readBooleanLoose(
  data: Record<string, unknown>,
  key: string,
  fallback: boolean,
): boolean {
  const value = data[key];
  if (typeof value === 'boolean') {
    return value;
  }
  if (typeof value === 'number') {
    return value !== 0;
  }
  if (typeof value === 'string') {
    const normalized = value.trim().toLowerCase();
    if (['1', 'true', 'yes', 'on'].includes(normalized)) {
      return true;
    }
    if (['0', 'false', 'no', 'off'].includes(normalized)) {
      return false;
    }
  }
  return fallback;
}

function isValidTime(value: string): boolean {
  if (!timeRegExp.test(value)) {
    return false;
  }
  const hour = Number(value.slice(0, 2));
  const minute = Number(value.slice(3, 5));
  return hour >= 0 && hour <= 23 && minute >= 0 && minute <= 59;
}

function resolveDate(raw: string): string {
  if (raw) {
    return raw;
  }
  return new Date().toISOString().slice(0, 10);
}

function toTrimmedString(value: unknown): string {
  return String(value ?? '').trim();
}

function normalizeMedicineRefs(raw: unknown): ReminderMedicineRefLike[] {
  if (!Array.isArray(raw)) {
    return [];
  }

  const normalized: ReminderMedicineRefLike[] = [];
  const dedupe = new Set<string>();
  for (const item of raw) {
    if (!item || typeof item !== 'object') {
      continue;
    }
    const record = item as Record<string, unknown>;
    const productName = toTrimmedString(record.productName);
    if (!productName) {
      continue;
    }
    const drugCode = toTrimmedString(record.drugCode);
    const approvalNo = toTrimmedString(record.approvalNo);
    const key = `${drugCode}|${approvalNo}|${productName}`;
    if (dedupe.has(key)) {
      continue;
    }
    dedupe.add(key);
    normalized.push({ drugCode, approvalNo, productName });
  }
  return normalized;
}

function deriveMedicinesFromLegacy(
  drugCode: string,
  approvalNo: string,
  productName: string,
): ReminderMedicineRefLike[] {
  const name = productName.trim();
  if (!name) {
    return [];
  }
  return [
    {
      drugCode: drugCode.trim(),
      approvalNo: approvalNo.trim(),
      productName: name,
    },
  ];
}

function composeProductName(
  medicines: ReminderMedicineRefLike[],
  fallback: string,
): string {
  const names = medicines
    .map(item => item.productName.trim())
    .filter(Boolean);
  const uniqueNames = [...new Set(names)];
  if (uniqueNames.length > 0) {
    return uniqueNames.join('、');
  }
  return fallback.trim();
}

function normalizeMedicinesFromDoc(doc: ReminderDocLike): ReminderMedicineRefLike[] {
  const medicines = normalizeMedicineRefs(doc.medicines);
  if (medicines.length > 0) {
    return medicines;
  }
  return deriveMedicinesFromLegacy(
    String(doc.drugCode ?? '').trim(),
    String(doc.approvalNo ?? '').trim(),
    String(doc.productName ?? '').trim(),
  );
}

function pickPrimaryIdentity(
  medicines: ReminderMedicineRefLike[],
  fallbackDrugCode: string,
  fallbackApprovalNo: string,
): { drugCode: string; approvalNo: string } {
  let primary: ReminderMedicineRefLike | null = null;
  for (const item of medicines) {
    if (item.drugCode.trim() || item.approvalNo.trim()) {
      primary = item;
      break;
    }
  }
  primary ??= medicines.length > 0
    ? medicines[0]
    : {
        drugCode: fallbackDrugCode,
        approvalNo: fallbackApprovalNo,
        productName: '',
      };

  return {
    drugCode: primary.drugCode.trim().length > 0
        ? primary.drugCode.trim()
        : fallbackDrugCode,
    approvalNo: primary.approvalNo.trim().length > 0
        ? primary.approvalNo.trim()
        : fallbackApprovalNo,
  };
}

function isPlanActiveOnDate(doc: ReminderDocLike, date: string): boolean {
  const startDate = String(doc.startDate ?? '').trim();
  const endDate = String(doc.endDate ?? '').trim();
  if (startDate && date < startDate) {
    return false;
  }
  if (endDate && date > endDate) {
    return false;
  }
  return true;
}

function toPlanRecord(doc: ReminderDocLike): ReminderPlanRecord {
  const medicines = normalizeMedicinesFromDoc(doc);
  const legacyDrugCode = String(doc.drugCode ?? '').trim();
  const legacyApprovalNo = String(doc.approvalNo ?? '').trim();
  const primaryIdentity = pickPrimaryIdentity(
    medicines,
    legacyDrugCode,
    legacyApprovalNo,
  );
  const productName = composeProductName(
    medicines,
    String(doc.productName ?? '').trim(),
  );
  return {
    id: String(doc._id ?? '').trim(),
    userId: String(doc.userId ?? '').trim(),
    time: String(doc.time ?? '').trim(),
    drugCode: primaryIdentity.drugCode,
    approvalNo: primaryIdentity.approvalNo,
    productName,
    medicines,
    dosage: String(doc.dosage ?? '').trim(),
    subtitle: String(doc.subtitle ?? '').trim(),
    enabled: doc.enabled !== false,
    repeatRule: String(doc.repeatRule ?? 'daily').trim() || 'daily',
    method: String(doc.method ?? 'notification').trim() || 'notification',
    startDate: String(doc.startDate ?? '').trim(),
    endDate: String(doc.endDate ?? '').trim(),
  };
}

function toTodayItem(doc: ReminderDocLike): TodayReminderItemRecord {
  const plan = toPlanRecord(doc);
  return {
    id: String(doc._id ?? '').trim(),
    time: String(doc.time ?? '').trim(),
    title: plan.productName || '用药提醒',
    dosage: String(doc.dosage ?? '').trim(),
    subtitle: String(doc.subtitle ?? '').trim(),
    done: false,
  };
}

export async function handleReminderUpsert(
  body: unknown,
): Promise<ApiEnvelope<ReminderPlanRecord>> {
  try {
    const data = expectRecord(body);
    const userId = readTrimmedString(data, 'userId');
    const id = readTrimmedString(data, 'id');
    const time = readTrimmedString(data, 'time');
    const legacyProductName = readTrimmedString(data, 'productName');
    const legacyDrugCode = readTrimmedString(data, 'drugCode');
    const legacyApprovalNo = readTrimmedString(data, 'approvalNo');
    const medicinesFromBody = normalizeMedicineRefs(data['medicines']);
    const medicines = medicinesFromBody.length > 0
        ? medicinesFromBody
        : deriveMedicinesFromLegacy(
            legacyDrugCode,
            legacyApprovalNo,
            legacyProductName,
          );
    const productName = composeProductName(medicines, legacyProductName);
    const primaryIdentity = pickPrimaryIdentity(
      medicines,
      legacyDrugCode,
      legacyApprovalNo,
    );

    if (!userId) {
      return fail('userId 不能为空');
    }
    if (!time || !isValidTime(time)) {
      return fail('time 格式错误，应为 HH:mm');
    }
    if (!productName) {
      return fail('productName 不能为空');
    }

    const drugCode = primaryIdentity.drugCode;
    const approvalNo = primaryIdentity.approvalNo;
    const dosage = readTrimmedString(data, 'dosage');
    const subtitle = readTrimmedString(data, 'subtitle');
    const enabled = readBooleanLoose(data, 'enabled', true);
    const repeatRule = readTrimmedString(data, 'repeatRule', 'daily') || 'daily';
    const method =
      readTrimmedString(data, 'method', 'notification') || 'notification';
    const startDate = readTrimmedString(data, 'startDate');
    const endDate = readTrimmedString(data, 'endDate');

    if (startDate && !dateRegExp.test(startDate)) {
      return fail('startDate 格式错误，应为 YYYY-MM-DD');
    }
    if (endDate && !dateRegExp.test(endDate)) {
      return fail('endDate 格式错误，应为 YYYY-MM-DD');
    }
    if (startDate && endDate && startDate > endDate) {
      return fail('startDate 不能晚于 endDate');
    }

    const now = Date.now();

    if (id) {
      if (!isValidObjectId(id)) {
        return fail('id 格式错误');
      }

      const existing = await Reminder.findOne({ _id: id, userId }).lean();
      if (!existing) {
        return fail('提醒不存在');
      }

      await Reminder.updateOne(
        { _id: id, userId },
        {
          $set: {
            time,
            drugCode,
            approvalNo,
            productName,
            medicines,
            dosage,
            subtitle,
            enabled,
            repeatRule,
            method,
            startDate,
            endDate,
            updatedAt: now,
          },
        },
      );

      return success(
        toPlanRecord({
          ...existing,
          time,
          drugCode,
          approvalNo,
          productName,
          medicines,
          dosage,
          subtitle,
          enabled,
          repeatRule,
          method,
          startDate,
          endDate,
        }),
      );
    }

    const created = await Reminder.create({
      userId,
      time,
      drugCode,
      approvalNo,
      productName,
      medicines,
      dosage,
      subtitle,
      enabled,
      repeatRule,
      method,
      startDate,
      endDate,
      createdAt: now,
      updatedAt: now,
    });

    return success(toPlanRecord(created.toObject()));
  } catch (error) {
    console.error('reminder-upsert failed:', error);
    return toApiFailure(error, '保存提醒失败，请稍后重试');
  }
}

export async function handleReminderDelete(
  body: unknown,
): Promise<ApiEnvelope<boolean>> {
  try {
    const data = expectRecord(body);
    const userId = readTrimmedString(data, 'userId');
    const id = readTrimmedString(data, 'id');

    if (!userId) {
      return fail('userId 不能为空');
    }
    if (!id) {
      return fail('id 不能为空');
    }
    if (!isValidObjectId(id)) {
      return fail('id 格式错误');
    }

    const result = await Reminder.deleteOne({ _id: id, userId });
    return success(result.deletedCount === 1);
  } catch (error) {
    console.error('reminder-delete failed:', error);
    return toApiFailure(error, '删除提醒失败，请稍后重试');
  }
}

export async function handleReminderList(
  body: unknown,
): Promise<ApiEnvelope<ReminderListPayload>> {
  try {
    const data = expectRecord(body);
    const userId = readTrimmedString(data, 'userId');
    if (!userId) {
      return fail('userId 不能为空');
    }

    const rows = await Reminder.find({ userId })
      .sort({ time: 1, createdAt: 1 })
      .lean();

    return success({ items: rows.map(item => toPlanRecord(item)) });
  } catch (error) {
    console.error('reminder-list failed:', error);
    return toApiFailure(error, '查询提醒列表失败，请稍后重试');
  }
}

export async function handleTodayReminders(
  body: unknown,
): Promise<ApiEnvelope<TodayRemindersPayload>> {
  try {
    const data = expectRecord(body);
    const dateInput = readTrimmedString(data, 'date');
    if (dateInput && !dateRegExp.test(dateInput)) {
      return fail('date 格式错误，应为 YYYY-MM-DD');
    }

    const date = resolveDate(dateInput);
    const userId = readTrimmedString(data, 'userId');
    if (!userId) {
      return success({ date, items: [] });
    }

    const rows = await Reminder.find({
      userId,
      enabled: { $ne: false },
    })
      .sort({ time: 1, createdAt: 1 })
      .lean();

    return success({
      date,
      items: rows.filter(item => isPlanActiveOnDate(item, date)).map(item => toTodayItem(item)),
    });
  } catch (error) {
    console.error('today-reminders failed:', error);
    return toApiFailure(error, '查询今日提醒失败，请稍后重试');
  }
}

File: lib/features/reminders/presentation/models/reminder.dart
import 'package:json_annotation/json_annotation.dart';

import 'package:luminous/utils/app_i18n_text.dart';

part 'reminder.g.dart';

/// 提醒中绑定的单个药品引用。
@JsonSerializable(createFactory: false)
class ReminderMedicineRef {
  final String drugCode;
  final String approvalNo;
  final String productName;

  const ReminderMedicineRef({
    this.drugCode = '',
    this.approvalNo = '',
    required this.productName,
  });

  factory ReminderMedicineRef.fromJson(Map<String, dynamic> json) {
    return ReminderMedicineRef(
      drugCode: (json['drugCode'] ?? '').toString().trim(),
      approvalNo: (json['approvalNo'] ?? '').toString().trim(),
      productName: (json['productName'] ?? '').toString().trim(),
    );
  }

  Map<String, dynamic> toJson() => _$ReminderMedicineRefToJson(this);
}

/// 用药提醒计划对象。
///
/// 该对象既用于：
/// - 后端接口的序列化/反序列化；
/// - 本地 SQLite 缓存（reminders 表）。
@JsonSerializable(createFactory: false)
class ReminderPlan {
  /// 提醒计划 id（后端可能返回 `id` 或 `_id`）。
  final String id;

  /// 所属用户 id。
  final String userId;

  /// 提醒时间（HH:mm）。
  final String time;

  /// 药品编码（可选）。
  final String drugCode;

  /// 批准文号（可选）。
  final String approvalNo;

  /// 药品名称（用于通知标题/列表标题）。
  final String productName;

  /// 当前提醒绑定的药品列表。
  final List<ReminderMedicineRef> medicines;

  /// 服用剂量（例如 1 粒 / 5 ml）。
  final String dosage;

  /// 提醒副标题（例如“早餐后服用 1 粒”）。
  final String subtitle;

  /// 是否启用该提醒。
  final bool enabled;

  /// 重复规则（当前主要支持 daily）。
  final String repeatRule;

  /// 提醒方式（当前主要支持 notification）。
  final String method;

  /// 生效开始日期（yyyy-MM-dd，留空表示不限制）。
  final String startDate;

  /// 生效结束日期（yyyy-MM-dd，留空表示不限制）。
  final String endDate;

  /// 创建一个提醒计划对象。
  const ReminderPlan({
    required this.id,
    required this.userId,
    required this.time,
    required this.drugCode,
    required this.approvalNo,
    required this.productName,
    this.medicines = const [],
    this.dosage = '',
    required this.subtitle,
    required this.enabled,
    required this.repeatRule,
    required this.method,
    this.startDate = '',
    this.endDate = '',
  });

  /// 从后端 JSON 反序列化为 `ReminderPlan`。
  factory ReminderPlan.fromJson(Map<String, dynamic> json) {
    final legacyDrugCode = (json['drugCode'] ?? '').toString().trim();
    final legacyApprovalNo = (json['approvalNo'] ?? '').toString().trim();
    final legacyProductName = (json['productName'] ?? '').toString().trim();
    final medicines = _parseMedicines(
      json['medicines'],
      fallbackDrugCode: legacyDrugCode,
      fallbackApprovalNo: legacyApprovalNo,
      fallbackProductName: legacyProductName,
    );
    final productName = legacyProductName.isNotEmpty
        ? legacyProductName
        : _composeProductName(medicines);
    final primary = medicines.isNotEmpty ? medicines.first : null;

    return ReminderPlan(
      id: (json['id'] ?? json['_id'] ?? '').toString(),
      userId: (json['userId'] ?? '').toString(),
      time: (json['time'] ?? '').toString(),
      drugCode: legacyDrugCode.isNotEmpty
          ? legacyDrugCode
          : (primary?.drugCode ?? ''),
      approvalNo: legacyApprovalNo.isNotEmpty
          ? legacyApprovalNo
          : (primary?.approvalNo ?? ''),
      productName: productName,
      medicines: medicines,
      dosage: (json['dosage'] ?? '').toString(),
      subtitle: (json['subtitle'] ?? '').toString(),
      enabled: json['enabled'] != false,
      repeatRule: (json['repeatRule'] ?? 'daily').toString(),
      method: (json['method'] ?? 'notification').toString(),
      startDate: (json['startDate'] ?? '').toString(),
      endDate: (json['endDate'] ?? '').toString(),
    );
  }

  Map<String, dynamic> toJson() => _$ReminderPlanToJson(this);

  /// 是否有有效 id。
  ///
  /// 在部分流程中可用它判断“这是新增还是更新”。
  @JsonKey(includeToJson: false)
  bool get hasId => id.trim().isNotEmpty;

  /// 页面展示用的标题（时间 + 药品名）。
  ///
  /// 例如：`08:30 维生素D`。
  @JsonKey(includeToJson: false)
  String get displayTitle {
    final t = time.trim();
    final name = productName.trim().isNotEmpty
        ? productName.trim()
        : _composeProductName(medicines);
    final n = name.isEmpty
        ? AppI18nText.pick(zh: '未知药品', en: 'Unknown medicine')
        : name;
    return t.isEmpty ? n : '$t $n';
  }
}

List<ReminderMedicineRef> _parseMedicines(
  dynamic raw, {
  required String fallbackDrugCode,
  required String fallbackApprovalNo,
  required String fallbackProductName,
}) {
  final medicines = <ReminderMedicineRef>[];
  final dedupe = <String>{};

  if (raw is List) {
    for (final item in raw) {
      if (item is! Map) {
        continue;
      }
      final medicine = ReminderMedicineRef.fromJson(
        item.cast<String, dynamic>(),
      );
      if (medicine.productName.trim().isEmpty) {
        continue;
      }
      final key =
          '${medicine.drugCode.trim()}|${medicine.approvalNo.trim()}|${medicine.productName.trim()}';
      if (dedupe.add(key)) {
        medicines.add(medicine);
      }
    }
  }

  if (medicines.isNotEmpty) {
    return medicines;
  }

  final fallbackName = fallbackProductName.trim();
  if (fallbackName.isEmpty) {
    return const [];
  }

  return <ReminderMedicineRef>[
    ReminderMedicineRef(
      drugCode: fallbackDrugCode.trim(),
      approvalNo: fallbackApprovalNo.trim(),
      productName: fallbackName,
    ),
  ];
}

String _composeProductName(List<ReminderMedicineRef> medicines) {
  final names = medicines
      .map((item) => item.productName.trim())
      .where((name) => name.isNotEmpty)
      .toList(growable: false);
  if (names.isEmpty) {
    return '';
  }
  return names.toSet().join('、');
}

/// 提醒计划列表接口返回模型。
@JsonSerializable(createFactory: false, createToJson: false)
class ReminderListResult {
  /// 当前用户的提醒计划列表。
  final List<ReminderPlan> items;

  /// 创建一个提醒列表结果对象。
  const ReminderListResult({required this.items});

  /// 从后端 JSON 反序列化为 `ReminderListResult`。
  factory ReminderListResult.fromJson(Map<String, dynamic> json) {
    /// 原始 items 字段，类型可能不稳定。
    final raw = json['items'];

    /// 解析后的提醒计划对象列表。
    final items = raw is List
        ? raw
              .whereType<Map>()
              .map((e) => ReminderPlan.fromJson(e.cast<String, dynamic>()))
              .toList()
        : <ReminderPlan>[];
    return ReminderListResult(items: items);
  }

  Map<String, dynamic> toJson() => <String, dynamic>{
        'items': items.map((e) => e.toJson()).toList(),
      };
}
